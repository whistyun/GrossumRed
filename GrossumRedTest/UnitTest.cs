using ApprovalTests;
using ApprovalTests.Namers;
using ApprovalTests.Reporters;
using GrossumRed;
using NUnit.Framework;
using System;
using System.CodeDom.Compiler;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

namespace GrossumRedTest
{
    [UseReporter(typeof(DiffReporter))]
    public class UnitTest
    {
        static UnitTest()
        {
            Approvals.RegisterDefaultNamerCreation(() => new Renamer("Out"));
        }

        [Test]
        public void Test1()
        {
            var source = Utils.LoadText("VmCls.vm.cs");
            var generated = CompileManager.CompileString(source, "test", errInfo => { });

            Approvals.Verify(generated);
        }

        [Test]
        public void Test2()
        {
            var source = Utils.LoadText("VmCls2.vm.cs");
            var generated = CompileManager.CompileString(source, "test", errInfo => { });

            Approvals.Verify(generated);
        }

        [Test]
        public void Test3()
        {
            var errors = new List<CompilerError>();

            var source = Utils.LoadText("VmCls3.vm.cs");
            var generated = CompileManager.CompileString(source, "test", errors.Add);

            Approvals.Verify(errors.AsString());
        }
    }


    static class ListExt
    {

        public static string AsString<T>(this List<T> list)
            => AsString(list, elm => elm.ToString());

        public static string AsString<T>(this List<T> list, Func<T, string> format )
        {
            var buf = new StringBuilder();
            buf.AppendLine("[");
            buf.Append("  ");
            buf.AppendLine(String.Join(",\r\n  ", list.Select(elm => format(elm))));
            buf.AppendLine("]");

            return buf.ToString();
        }
    }


    class Renamer : UnitTestFrameworkNamer
    {
        private string dir;

        public override string SourcePath
        {
            get => Path.Combine(base.SourcePath, dir);
        }

        public Renamer(string dir)
        {
            this.dir = dir;
        }
    }
}