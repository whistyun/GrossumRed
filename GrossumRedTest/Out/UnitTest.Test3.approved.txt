[
  test(24,1) : error GRD0006: The accessor already has some code. Grossum Red cannot insert a code. If you won't treat this accessor by GrossumRed, add '@ignore' to the head of this property.,
  test(29,1) : error GRD0006: The accessor already has some code. Grossum Red cannot insert a code. If you won't treat this accessor by GrossumRed, add '@ignore' to the head of this property.,
  test(37,1) : error GRD0004: setter is duplicated,
  test(39,1) : error GRD0003: getter is duplicated
]
